spim/CPU/mem.cpp

```diff
@@ -388,6 +392,10 @@ read_mem_byte(MIPSImage &img, mem_addr addr)
 reg_word
 read_mem_half(MIPSImage &img, mem_addr addr)
 {
+  std::optional<reg_word> custom_read = img.custom_memory_read(addr);
+  if (custom_read.has_value())
+    return custom_read.value();
+
```

Comment:
MMIO addresses should expect reading/writing a word. It definitely complicates the logic if we include this one function for all 3 different reads/writes.

What you can do is add a virtual `custom_memory_read_word()`, `custom_memory_read_half()`, `custom_memory_read_byte()`, ... same for the custom writes. It still allows for the functionality of doing custom writes for half words and bytes, but doesnt interfere with our logic of word-only MMIOs.